src/main/java/teammates/ui/webapi/FeedbackSessionClosingRemindersAction.java

```diff
@@ -13,16 +13,63 @@
 import teammates.common.util.EmailWrapper;
 import teammates.common.util.Logger;
 import teammates.common.util.RequestTracer;
+import teammates.storage.sqlentity.DeadlineExtension;
+import teammates.storage.sqlentity.FeedbackSession;
 
 /**
  * Cron job: schedules feedback session closing emails to be sent.
  */
-class FeedbackSessionClosingRemindersAction extends AdminOnlyAction {
+public class FeedbackSessionClosingRemindersAction extends AdminOnlyAction {
 
     private static final Logger log = Logger.getLogger();
 
     @Override
     public JsonResult execute() {
+        executeForDatastoreFeedbackSessions();
+
+        List<FeedbackSession> sessions = sqlLogic.getFeedbackSessionsClosingWithinTimeLimit();
+
+        for (FeedbackSession session : sessions) {
+            RequestTracer.checkRemainingTime();
+            List<EmailWrapper> emailsToBeSent = sqlEmailGenerator.generateFeedbackSessionClosingEmails(session);
+            try {
+                taskQueuer.scheduleEmailsForSending(emailsToBeSent);
+                session.setClosingSoonEmailSent(true);
+            } catch (Exception e) {
+                log.severe("Unexpected error", e);
+            }
+        }
+
+        executeForDatastoreExtendedDeadlines();
+
+        // Group deadline extensions by feedback sessions
+        Collection<List<DeadlineExtension>> groupedDeadlineExtensions =
+                sqlLogic.getDeadlineExtensionsPossiblyNeedingClosingEmail()
+                    .stream()
+                    .collect(Collectors.groupingBy(de -> de.getFeedbackSession()))
+                    .values();
+
+        for (var deadlineExtensions : groupedDeadlineExtensions) {
+            RequestTracer.checkRemainingTime();
+
+            FeedbackSession session = deadlineExtensions.get(0).getFeedbackSession();
+            if (!session.isClosingEmailEnabled()) {
+                continue;
+            }
+
+            List<EmailWrapper> emailsToBeSent = sqlEmailGenerator
+                    .generateFeedbackSessionClosingWithExtensionEmails(session, deadlineExtensions);
+            taskQueuer.scheduleEmailsForSending(emailsToBeSent);
+
+            for (var de : deadlineExtensions) {
+                de.setClosingSoonEmailSent(true);
+            }
+        }
+
+        return new JsonResult("Successful");
+    }
+
+    private void executeForDatastoreFeedbackSessions() {
         List<FeedbackSessionAttributes> sessions = logic.getFeedbackSessionsClosingWithinTimeLimit();
 
         for (FeedbackSessionAttributes session : sessions) {
```

Comment:
i think there might be a need for checking if the session has been migrated already

Reply:
Goode catch! 👍